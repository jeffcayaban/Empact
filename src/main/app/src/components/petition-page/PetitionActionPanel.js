import React from 'react'
import { Button, Tooltip, OverlayTrigger, DropdownButton, MenuItem } from "react-bootstrap";

import { Route, withRouter } from "react-router-dom";
import { closePetition } from "../../utils/AppUtils";
import { generateGeneralConfirmDialog, generateSomethingWentWrongDialog, generateSuccessDialog } from "../../utils/Helpers";

/**
 * Provides the user with actions to be applied to a petition. This includes deleting, editing, or closing a
 * petition.
 */

class PetitionActionPanel extends React.PureComponent {

    // Handles the result of the delete petition request.
    handleDelete() {
        generateGeneralConfirmDialog().then((result) => {
            if (result.value) {
                // Assumes that the user has confirmed the action to delete the petition.

                this.props.deletePetition()
                    .then(response => {
                        generateSuccessDialog('Your petition has been deleted!', 'Explore').then(() => {
                            window.location = `/petitions`
                        })
                    }).catch(error => generateSomethingWentWrongDialog())
            }
        })
    }

    // Handles the result of the close petition request.
    handleClose(petitionId) {
        generateGeneralConfirmDialog().then((result) => {
            if (result.value) {
                // Assumes that the user has confirmed the action to close the petition.

                closePetition({ petitionId: petitionId })
                    .then(response => {
                        generateSuccessDialog('Your petition is now closed', 'Your Petition').then(() => {
                            window.location = `/petition/${petitionId}`
                        })
                    }).catch(error => generateSomethingWentWrongDialog())
            }
        })
    }

    // Generates the button to delete the petition.
    generateDeleteBtn(isOpen) {
        const deleteTooltip = <Tooltip id={"delete-petition-tooltip"}>Permanently delete this petition.</Tooltip>;
        const innerElement = isOpen
            ? <MenuItem onClick={this.handleDelete.bind(this)}>Delete</MenuItem>
            : <Button bsStyle="danger" onClick={this.handleDelete.bind(this)}>Delete</Button>;

        return (
            <OverlayTrigger placement={"top"} overlay={deleteTooltip}>
                { innerElement }
            </OverlayTrigger>
        )
    }

    render() {
        const editTooltip = <Tooltip id={"edit-petition-tooltip"}>Edit the contents of this petition.</Tooltip>;

        return (
            <div>
                <DropdownButton id={"petition-actions"} title={"Actions"} pullRight>
                    <Route render={({ history }) => (
                        <OverlayTrigger placement={"top"} overlay={editTooltip}>
                            <MenuItem onClick={() => { history.push(`/petition/edit/${this.props.petition.id}`) }}>
                                Edit
                            </MenuItem>
                        </OverlayTrigger>
                    )}/>
                    { this.generateDeleteBtn(true) }
                    <MenuItem divider />
                    <MenuItem onClick={() => this.handleClose(this.props.petition.id)}>Close</MenuItem>
                </DropdownButton>
            </div>
        );
    }
}

export default withRouter(PetitionActionPanel);