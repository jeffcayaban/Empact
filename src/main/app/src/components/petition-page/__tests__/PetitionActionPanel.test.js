import React from 'react';
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";

import MockPetition from '../../../other/mock-data/MockGetPetitionsResponse';

import PetitionActionPanel from "../PetitionActionPanel";

function createPetitionActionPanel(props = {}) {
    const mockRouter = jest.fn();
    return <PetitionActionPanel.WrappedComponent {...props} params={{ router: mockRouter }} />;
}

describe('PetitionActionPanel', () => {

    it('should render successfully given the petition is not closed', () => {
        const mockDeletePetition = () => Promise.resolve();

        const props = {
            isPetitionClosed: false,
            petition: MockPetition.content[0],
            deletePetition: mockDeletePetition
        };

        const component = shallow(createPetitionActionPanel(props));

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render successfully given the petition is closed', () => {
        const mockDeletePetition = () => Promise.resolve();

        const props = {
            isPetitionClosed: true,
            petition: MockPetition.content[0],
            deletePetition: mockDeletePetition
        };

        const component = shallow(createPetitionActionPanel(props));

        expect(toJson(component)).toMatchSnapshot();
    });

});
