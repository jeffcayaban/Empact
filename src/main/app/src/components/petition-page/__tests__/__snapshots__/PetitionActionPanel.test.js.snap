// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PetitionActionPanel should render successfully given the petition is closed 1`] = `
<div>
  <DropdownButton
    id="petition-actions"
    pullRight={true}
    title="Actions"
  >
    <Route
      render={[Function]}
    />
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="delete-petition-tooltip"
          placement="right"
        >
          Permanently delete this petition.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        header={false}
        onClick={[Function]}
      >
        Delete
      </MenuItem>
    </OverlayTrigger>
    <MenuItem
      bsClass="dropdown"
      disabled={false}
      divider={true}
      header={false}
    />
    <MenuItem
      bsClass="dropdown"
      disabled={false}
      divider={false}
      header={false}
      onClick={[Function]}
    >
      Close
    </MenuItem>
  </DropdownButton>
</div>
`;

exports[`PetitionActionPanel should render successfully given the petition is not closed 1`] = `
<div>
  <DropdownButton
    id="petition-actions"
    pullRight={true}
    title="Actions"
  >
    <Route
      render={[Function]}
    />
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="delete-petition-tooltip"
          placement="right"
        >
          Permanently delete this petition.
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        header={false}
        onClick={[Function]}
      >
        Delete
      </MenuItem>
    </OverlayTrigger>
    <MenuItem
      bsClass="dropdown"
      disabled={false}
      divider={true}
      header={false}
    />
    <MenuItem
      bsClass="dropdown"
      disabled={false}
      divider={false}
      header={false}
      onClick={[Function]}
    >
      Close
    </MenuItem>
  </DropdownButton>
</div>
`;
